src/main/java/org/jenkinsci/plugins/workflow/cps/CpsThreadGroup.java

```diff
@@ -200,8 +204,29 @@ private void assertVmThread() {
         assert current()==this;
     }
 
+    /**
+     * Returns the thread with the specified id.
+     *
+     * Normally called from the CPS VM thread, but may be called from other threads via {@link CpsStepContext#doGet}.
+     *
+     * @return
+     *      null if the thread has finished executing.
+     */
     public CpsThread getThread(int id) {
-        return threads.get(id);
+        CpsThread thread = threads.get(id);
+        if (thread == null && LOGGER.isLoggable(Level.FINE)) {
+            LOGGER.log(Level.FINE, "no thread " + id + " among " + threads.keySet(), new IllegalStateException());
```

Comment:
this is perhaps a little nitpicky, but...while this code is being touched, may as well disambiguate the error logging. that is, in the current situation, there's no way to distinguish between the cases where `id` is present in the map, but is mapped to a `null` value, and the case where `id` is simply absent.

two approaches come to mind:

a) use `Map::getOrDefault` with some kinda dummy default (...which would require _defining_ some dummy instance of `CpsThread`, which is likely overkill for this use case)
b) log `threads.entrySet()` rather than `threads.keySet()`
c) change the value type from `CpsThread` to `Optional<CpsThread>` :trollface: 

(also, I'd stress again that I am totally ignorant about this codebase, so if the possibility of a `null` value being inserted into the threads map is sufficiently remote as to be n/a, then never mind)

Reply:
Good point, but I think since we are using a standard `TreeMap` with natural ordering of keys, `null` is never permitted as a value and trying to insert `null` would throw an exception, see [TreeMap.put docs](https://docs.oracle.com/javase/8/docs/api/java/util/TreeMap.html#put-K-V-).

Reply:
ah, true. since `threads` is constructed without a comparator, it would use natural ordering, and thus disallow null.

in other words, _with that context_, the current logging is not ambiguous. if, as shown in some [sample output in the ticket](https://issues.jenkins-ci.org/browse/JENKINS-56890?focusedCommentId=364471&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-364471), `threads.get($id)` returns `null`, yet `$id` is present in the key set that is subsequently logged, then you know that there is a concurrency bug.

Reply:
that said, I would personally still log `threads.entrySet()`, unless there is some good reason not to (which seems quite unlikely).

Reply:
> I would personally still log threads.entrySet()

We could, but `CpsThread.toString` returns output like `Thread #id @hashcode`, so it probably wouldn't be particularly useful since the Map's keys are the ids.

Reply:
my reasoning in favor of `entrySet()` is that it would eliminate the potential _appearance_ of ambiguity vis-a-vis the possibility of a `null` value for anyone looking at the logging output without context.

for example, here is an excerpt from the comment I linked above:

> In both cases, the thread IDs being looked up do exist in the `threads` TreeSet, but their value is null? But I don't see anywhere where a null value gets inserted into that TreeSet. Could this be a race condition, accessing that value as it's being added/removed by another thread?

if `threads.entrySet()` were logged, this confusion wouldn't happen (even if the `toString` implementation of `CpsThread` is super minimal).

Reply:
Better to just use annotations/assertions to verify that null values could never be added.